### PR TITLE
Add Secrets Manager as a source for configuration values

### PR DESCRIPTION
**Pull request by mitchellhenke**
Similar to #40, but with a slightly different approach. This PR is in draft status for now to help move towards something that can accommodate the multiple use cases we may have for secrets manager usage in applications.

### CODE REVIEW

File: lib/identity/hostdata/config_builder.rb [hunk 1]
```diff
@@ -77,6 +73,28 @@ def add(key, type: :string, allow_nil: false, enum: nil, options: {})
...
+      def fetch_value_from_source(key, value)
+        if value.is_a?(Array) && value.length == 2 && value.first == 'env'
+          ENV.fetch(value[1])
+        elsif value.is_a?(Array) && value.length == 2 && value.first == 'secrets_manager'
+          secrets_client.get_secret_value(secret_id: value[1]).secret_string
+        elsif value.is_a?(Array)
+          raise "invalid configuration value for #{key}"
+        else
+          value
+        end
```
Comment (zachmargolis) [inline lib/identity/hostdata/config_builder.rb:85]: WDYT of:
1. adding hash support
2. DRYing the the array checks:
```suggestion
        if value.is_a?(Array)
          type, name, *rest = value
          case type
          when 'env'
            ENV.fetch(name)
          when 'secrets_manager'
            secrets_client.get_secret_value(secret_id: name).secret_string
          else
            raise "invalid configuration value for #{key}"
          end
        else
          value
        end
```

File: spec/identity/hostdata/config_builder_spec.rb [hunk 2]
```diff
@@ -74,6 +125,7 @@
         commas_key: :comma_separated_string_list,
         json_array: :json,
         string_env_key: :string,
+        :string_secrets_manager_key => :string,
```
Comment (zachmargolis) [inline spec/identity/hostdata/config_builder_spec.rb:128]:
```suggestion
        string_secrets_manager_key: :string,
```

File: lib/identity/hostdata/config_builder.rb (hunk 1 above)
Comment (mitchellhenke) [reply lib/identity/hostdata/config_builder.rb]: I think I'd rather tackle the hash separately from this to keep this focused on secrets manager support.

File: lib/identity/hostdata/config_builder.rb [hunk 3]
```diff
@@ -90,20 +84,47 @@ def add(
...
+
+      def secrets_client
+        @secrets_client ||= Aws::SecretsManager::Client.new(
+          region: Identity::Hostdata.aws_region,
+          http_idle_timeout: 5,
+          http_open_timeout: 5,
+          http_read_timeout: 5,
+          instance_profile_credentials_retries: 3,
+        )
+      end
```
Comment (zachmargolis) [inline lib/identity/hostdata/config_builder.rb:126]: this is a dupe now I think

File: identity-hostdata.gemspec [hunk 4]
```diff
@@ -39,4 +39,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_development_dependency "rspec", "~> 3.0"
   spec.add_development_dependency "webmock", "~> 3.0.1"
+  spec.add_development_dependency "rexml"
```
Comment (zachmargolis) [inline identity-hostdata.gemspec:42]: is this for some other dependency?

Comment (mitchellhenke) [reply identity-hostdata.gemspec:42]: Without it, I get the following error:

```
An error occurred while loading ./spec/identity/hostdata_spec.rb.
Failure/Error: require 'aws-sdk-secretsmanager'

RuntimeError:
  Unable to find a compatible xml library. Ensure that you have installed or added to your Gemfile one of ox, oga, libxml, nokogiri or rexml
```

**Review by zachmargolis — APPROVED** [review]

**Pull request merged**